### PR TITLE
eth status

### LINKED ISSUE #175: Add cli status commands
## Summary of bug
 
One cannot tell if the status of the eth connection nor of plasmad
## Expected behavior

Easily figure out the current state of the eth connection (syncing, crashed etc) and of plasmad (current block num)

Add `status` commands to plasmacli, perferably `plasmacli eth status` and `plasmacli query status`

### PR DESCRIPTION
**Pull request by wesgraham**
Fixes #175 
eth status for plasmacli. Returns connection status to eth node

### CODE REVIEW

**Comment by hamdiallam** [comment]
The `.idea` editor files and geth.service script was committed. Shouldn't be apart of the PR. looks good though

File: cmd/plasmacli/subcmd/eth/status.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		client, err := eth.InitEthConn(conf.EthNodeURL)
+		if err != nil {
+			return fmt.Errorf("error initializing connection with client: { %s }", err)
+		}
+
+		// Check whether client synced
+		synced, err := client.Synced()
+		if err != nil {
+			return fmt.Errorf("error checking synced status: { %s }", err)
+		}
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/eth/status.go:37]: Past this, I would not consider the following statement errors. Should switch these to normal print statements. Also you don't need to camel case statements that are meant to be consumed by the reader, "ethNode".

i.e
if (synced) print("live and synced full node.")
else print("live full node. The node is not fully synced.")

File: cmd/plasmacli/subcmd/query/root.go [hunk 2]
```diff
@@ -13,6 +13,7 @@ func RootCmd() *cobra.Command {
 		BlockCmd(),
 		BlocksCmd(),
 		InfoCmd(),
+		//StatusCmd(),
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/query/root.go]: Is this meant to be commented out?

File: cmd/plasmacli/subcmd/eth/root.go [hunk 3]
```diff
@@ -40,6 +40,7 @@ func RootCmd() *cobra.Command {
 		FinalizeCmd(),
 		DepositCmd(),
 		WithdrawCmd(),
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/eth/root.go:42]: Lets add a line break under `WithdrawCmd`. and have the status be on a new line.

Separates the cmds that are action based versus a health check

File: cmd/plasmacli/subcmd/query/root.go (hunk 2 above)
Comment (hamdiallam) [reply cmd/plasmacli/subcmd/query/root.go]: Don't think we need this here actually.

**Review by hamdiallam — CHANGES_REQUESTED** [review]

File: cmd/plasmacli/main.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ import (
 	"os"
 )
 
+
```
Comment (hamdiallam) [inline cmd/plasmacli/main.go:9]: remove this line

File: cmd/plasmacli/subcmd/query/status.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+// StatusCmd returns the current state of the eth connection (syncing, crashed etc)
+func StatusCmd() *cobra.Command {
+	return statuscmd
+}
+
+var statuscmd = &cobra.Command {
+	Use:   "status",
+	Short: "Check current block number",
+	Long: `returns current block number of plasmad connection`,
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/query/status.go]: s/`/" and s/returns/Returns

Do you think status makes here as the command name? wdyt about "height"? `plasmacli query height`

File: cmd/plasmacli/subcmd/query/status.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package query
+
+import (
+	"fmt"
+	"github.com/FourthState/plasma-mvp-sidechain/client"
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/spf13/cobra"
+)
+
+// StatusCmd returns the current state of the eth connection (syncing, crashed etc)
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/query/status.go]: Not the right godoc. Think this godoc is meant for the other status command

**Review by hamdiallam — COMMENTED** [review]
Nice, almost there.

File: cmd/plasmacli/subcmd/eth/status.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	Use:   "status",
+	Short: "check state of eth connection",
+	Long: `returns current state of eth connection (syncing, crashed, etc)`,
+	SilenceUsage: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		// Get Conf
+		conf, err := config.ParseConfigFromViper()
+		if err != nil {
+			return fmt.Errorf("error retrieving config: { %s }", err)
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/eth/status.go]: also, can you remove the `{` wrapper. Common go error format is "err: suberr". I know this is inconsistent in other parts of the codebase but working towards fixing it

Comment (hamdiallam) [reply cmd/plasmacli/subcmd/eth/status.go]: same applies for errors below

File: .gitignore [hunk 8]
```diff
@@ -4,3 +4,4 @@ contracts/abi
 contracts/node_modules
 cmd/plasmad/plasmad
 cmd/plasmacli/plasmacli
+/.idea
```
Comment (hamdiallam) [inline .gitignore]: should be `./idea`

File: cmd/plasmacli/subcmd/query/status.go (hunk 5 above)
Comment (colin-axner) [reply cmd/plasmacli/subcmd/query/status.go]: if `status` is used, I think we should maybe print out more info of the connected client

File: cmd/plasmacli/subcmd/eth/status.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+		}
+
+		// Initialize Eth Connection
+		client, err := eth.InitEthConn(conf.EthNodeURL)
+		if err != nil {
+			return fmt.Errorf("error initializing connection with client: { %s }", err)
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/eth/status.go]: I think the `ParseConfigFromViper()` and `InitEthConn` can be replaced with a call to [config.GetContractConn()](https://github.com/FourthState/plasma-mvp-sidechain/blob/536e4ac2156bfa6add7fc76da3ce123820437fc8/cmd/plasmacli/config/connection.go#L14)

It does the same thing (DRY) so might as well use the library and the library has verbose error messages.

File: cmd/plasmacli/subcmd/eth/status.go (hunk 1 above)
Comment (colin-axner) [reply cmd/plasmacli/subcmd/eth/status.go:37]: should eth block height be printed to?

**Review by colin-axner — CHANGES_REQUESTED** [review]
Great work! Thanks for handling this! left some feedback in `subcmd/eth/status.go`, let me know if you have any questions, `eth/root.go` has example usage for what I mentioned

File: cmd/plasmacli/subcmd/eth/status.go (hunk 9 above)
Comment (hamdiallam) [reply cmd/plasmacli/subcmd/eth/status.go]: Ah good catch, that's right

Comment (hamdiallam) [reply cmd/plasmacli/subcmd/eth/status.go]: Ah wait, `config.GetContractConn` provides the `eth.Plasma` struct, in which the ethereum client is a private field. 

Is it worth trying to improve this?

File: cmd/plasmacli/subcmd/query/status.go (hunk 5 above)
Comment (wesgraham) [reply cmd/plasmacli/subcmd/query/status.go]: Lets go with `height` for now. Can open up another issue with other metrics we are interested in (and eventually combine). Fixing now

File: cmd/plasmacli/subcmd/eth/status.go (hunk 9 above)
Comment (wesgraham) [reply cmd/plasmacli/subcmd/eth/status.go]: @colin-axner will leave it to you to make the call!

File: cmd/plasmacli/subcmd/query/status.go (hunk 6 above)
Comment (wesgraham) [reply cmd/plasmacli/subcmd/query/status.go]: What should it be?

Comment (hamdiallam) [reply cmd/plasmacli/subcmd/query/status.go]: this checks the current block number

File: cmd/plasmacli/subcmd/eth/status.go (hunk 9 above)
Comment (hamdiallam) [reply cmd/plasmacli/subcmd/eth/status.go]: Let's leave it how you have it for now for the reason I listed. Unless you have other opinions @colin-axner.

File: cmd/plasmacli/subcmd/query/height.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+// HeightCmd returns the current block height of the plasmad connection
+func HeightCmd() *cobra.Command {
+	return heightcmd
+}
+
+var heightcmd = &cobra.Command {
+	Use:   "height",
+	Short: "check current block height",
+	Long: "returns current block height of plasmad connection",
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/query/height.go]: s/returns/Returns, s/check/Check

File: cmd/plasmacli/subcmd/query/height.go [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	SilenceUsage: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		ctx := context.NewCLIContext()
+		height, err := client.Height(ctx)
+		if err != nil {
+			return fmt.Errorf("error retrieving current block height: %s", err)
+		} else {
```
Comment (hamdiallam) [inline cmd/plasmacli/subcmd/query/height.go]: nit: go style

Don't need the `else` block here because there is no possible exit out first conditional (it returns the error)

**Review by hamdiallam — APPROVED** [review]
LGTM.

I added a few comments. Make sure to select Squash & Merge option

File: .idea/.gitignore [hunk 12]
```diff
@@ -0,0 +1,3 @@
+
+# Default ignored files
+/workspace.xml
```
Comment (colin-axner) [inline .idea/.gitignore:3]: why is this file necessary?

File: cmd/plasmacli/subcmd/query/height.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+// HeightCmd returns the current block height of the plasmad connection
+func HeightCmd() *cobra.Command {
+	return heightcmd
+}
+
+var heightcmd = &cobra.Command {
+	Use:   "height",
+	Short: "Check current block height",
+	Long: "Returns current block height of plasmad connection",
+	SilenceUsage: true,
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/query/height.go:19]: have all these files used `gofmt` I feel like the spacing is off

File: cmd/plasmacli/subcmd/eth/status.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+		}
+		if synced {
+			num, err := client.LatestBlockNum()
+			if err != nil {
+				return fmt.Errorf("synced with eth node: %s \n error retrieving latest block height of eth endpoint: %s\n", conf.EthNodeURL , err)
+			}
+			fmt.Printf("synced with eth node: %s \nlatest block height of the eth endpoint: %d\n", conf.EthNodeURL, num)
+
+		} else {
+			fmt.Printf("could not sync with eth node: %s", conf.EthNodeURL)
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/eth/status.go:46]: I think this message should say something like `"eth node is syncing: %s", conf.EthNodeURL`. The connection isn't an issue by this point, if synced == false it's because it's still trying to catch up to the latest height

**Review by colin-axner — CHANGES_REQUESTED** [review]

File: cmd/plasmacli/subcmd/eth/status.go (hunk 9 above)
Comment (colin-axner) [reply cmd/plasmacli/subcmd/eth/status.go]: I'm fine with an issue being opened to do these changes later, code can be improved but functionality wouldn't change so not a blocking request

File: .idea/.gitignore (hunk 12 above)
Comment (hamdiallam) [reply .idea/.gitignore:3]: ah yeah, can you remove the `.idea` folder?

File: cmd/plasmacli/main.go (hunk 4 above)
Comment (hamdiallam) [inline cmd/plasmacli/main.go:9]: Can you remove this line

**Review by hamdiallam — CHANGES_REQUESTED** [review]

File: cmd/plasmacli/subcmd/query/height.go (hunk 13 above)
Comment (wesgraham) [reply cmd/plasmacli/subcmd/query/height.go:19]: just fixed

**Comment by wesgraham** [comment]
Made a new branch for this issue so I could work on "export" in parallel, opening a new PR to reflect recent changes
